pkg/controllers/nodeclaim/capacityreservation/controller.go

```diff
@@ -0,0 +1,83 @@
+package capacityreservation
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/awslabs/operatorpkg/singleton"
+	"github.com/samber/lo"
+	"go.uber.org/multierr"
+	"k8s.io/apimachinery/pkg/api/errors"
+	controllerruntime "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	karpv1 "sigs.k8s.io/karpenter/pkg/apis/v1"
+	"sigs.k8s.io/karpenter/pkg/cloudprovider"
+	"sigs.k8s.io/karpenter/pkg/operator/injection"
+)
+
+type Controller struct {
+	cp cloudprovider.CloudProvider
+	kubeClient client.Client
+}
+
+func NewController() *Controller {
+	return nil
+}
+
+func (*Controller) Name() string {
+	return "nodeclaim.capacityreservation"
+}
+
+func (c *Controller) Register(_ context.Context, m manager.Manager) error {
+	return controllerruntime.NewControllerManagedBy(m).
+		Named(c.Name()).
+		WatchesRawSource(singleton.Source()).
+		Complete(singleton.AsReconciler(c))
+}
+
+func (c *Controller) Reconcile(ctx context.Context) (reconcile.Result, error) {
+	ctx = injection.WithControllerName(ctx, c.Name())
+
+	cpNodeClaims, err := c.cp.List(ctx)
+	if err != nil {
+		return reconcile.Result{}, fmt.Errorf("listing instance types, %w", err)
+	}
+	cpNodeClaimIndex := lo.SliceToMap(cpNodeClaims, func(nc *karpv1.NodeClaim) (string, *karpv1.NodeClaim) {
+		return nc.Status.ProviderID, nc
+	})
+
+
+	var ncs *karpv1.NodeClaimList
+	if err := c.kubeClient.List(ctx, ncs); err != nil {
+		return reconcile.Result{}, fmt.Errorf("listing nodeclaims, %w", err)
+	}
+	var errs []error
+	for i := range ncs.Items {
```

Comment:
```suggestion
	for _, nc := range ncs.Items {
```

Can we use the value of the NodeClaim rather than indexing now that we have the pointer issue that was previously an issue fixed in 1.23

Reply:
That's not actually why I was using an index, I was just using it to avoid making copies of the whole list. Probably doesn't matter too much here since we're not in a hot path, but I'd rather leave it as is.

Reply:
If you don't pull the whole list, then I think it doesn't even matter anymore